### ISSUE TITLE
thread_rng can't get random number in a WASI module

### ISSUE THREAD

**Issue by autodidaddict** [wasmerio/wasmer#460]
### Runtime panic (trap) when getting random number

Rust code in the compiled WASI module:  `let mut rng = thread_rng();`

My system:
```
wasmer 0.4.2 | rustc 1.34.2 (6c2484dc3 2019-05-13) | x86_64
```

### Steps to reproduce

1. Create a Rust project
2. Have that project invoke `thread_rng()` to produce a RNG
3. Compile Rust target to WASI
4. attempt to run module with wasmer

### Expected behavior

According to the WASI spec, there's a __wasi_rnd_get() function that I figured would be invoked underneath the `rand` crate's compiled WASI code.

### Actual behavior

```
thread 'main' panicked at 'could not initialize thread_rng: All entropy sources failed (permanently unavailable); cause: OsRng: support for wasm32 requires emscripten, stdweb or wasm-bindgen (permanently unavailable)', /home/kevin/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.6.5/src/rngs/thread.rs:80:17
```

This error message reads like someone's compile-time error message that doesn't take into account the shunting being done by the WASI code and its relevant imports.

If we can't use random numbers, things are going to be difficult compiling to WASI.

### Additional context

I'm perfectly fine with cheap workarounds. I don't need cryptographic random, I'm just using this to shuffle a list of servers to connect with, so if there are some lines of Rust I can replace my code with that will pick a random number between 0 and n or let me invoke `shuffle` on a vec and not cause a wasmer-time trap, that'd be awesome.

p.s. I don't know if the problem is in the runtime wasmer's providing or in the code that Rust is generating via the `wasm32-wasi` target. Apologies if this is not your issue at all.

**Comment by autodidaddict** [comment]
I've completely removed all random number generation from my own code, but it looks like somewhere in the dependency tree there's still some other code using `thread_rng` from the `rand` crate.

**Comment by Hywan** [comment]
Did you consider using https://github.com/rust-random/rand to get random number with WebAssembly?

**Comment by autodidaddict** [comment]
@Hywan yeah, like I said I removed my own random code. Problem is the use of the rand crate is littered throughout my dependency tree.

**Comment by Hywan** [comment]
Try to run `cargo update` or even `cargo clean` at some point.

**Comment by autodidaddict** [comment]
I've done both. I can actually _compile_ the wasm to the `wasm32-wasi` target. It just panics as soon as `wasmer` 's execution hits a spot that tries to invoke `thread_rng()`

**Comment by MarkMcCaskey** [comment]
Thanks for the report!  I'll investigate and write a test covering this case

**Comment by MarkMcCaskey** [comment]
The error message is originating past the rand crate.

See https://github.com/rust-random/rand/issues/769 for more information.  
> And no, we don't have an ETA on the Rand 0.7 release. See #715 and the milestone. Ideally it will be within a couple of months.

That was early April, so hopefully WASI support will be coming soon!

**Comment by huangjj27** [comment]
Encountering the same issue.

**Comment by jedisct1** [comment]
You can use the `getrandom` crate instead.

**Comment by kubkon** [comment]
> The error message is originating past the rand crate.
> 
> See [rust-random/rand#769](https://github.com/rust-random/rand/issues/769) for more information.
> 
> > And no, we don't have an ETA on the Rand 0.7 release. See #715 and the milestone. Ideally it will be within a couple of months.
> 
> That was early April, so hopefully WASI support will be coming soon!

Hey guys! I don't know whether this issue is still relevant, but just wanted to let you know that the prelease version `0.7.0-pre.0` of `rust-random/rand` comes with `wasm32-wasi` support :-) So you can just specify it in your `Cargo.toml` as:

```
[dependencies]                                                                                                          
rand = "0.7.0-pre.0"
```

**Comment by huangjj27** [comment]
> Hey guys! I don't know whether this issue is still relevant, but just wanted to let you know that the prelease version 0.7.0-pre.0 of rust-random/rand comes with wasm32-wasi support

Error has changed. It seems that is some issue with wasmer support:
`thread 'main' panicked at 'wasm trap occured: unreachable', src\bin\wasmer.rs:515:51`

**Comment by syrusakbary** [comment]
This is fixed in the new version of `rand`: `0.7`
https://github.com/rust-random/rand/issues/715

Closing the issue
